### PR TITLE
drivers/mm: Add API to query memory regions

### PR DESCRIPTION
**Pull request by edersondisouza**
Some systems may need to query the available memory regions in runtime.
For those, this patch adds a simple API that memory management drivers
can implement to provide such discovery.

A small test also added to exercise the default, empty implementation.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

### CODE REVIEW

File: include/zephyr/drivers/mm/system_mm.h [hunk 1]
```diff
@@ -339,6 +339,48 @@ int sys_mm_drv_update_page_flags(void *virt, uint32_t flags);
...
+	for (iter = regions; iter->size; iter++)
+
+/**
+ * @brief Query available memory regions
+ *
+ * Returns an array of available memory regions. One can iterate over
+ * the array using #SYS_MM_DRV_MEMORY_REGION_FOREACH. Note that the last
+ * item of the array is identified by it's size attribute, which is zero.
+ *
+ * @retval regions A possibly empty array of memory regions.
```
Comment (marcinszkudlinski) [inline include/zephyr/drivers/mm/system_mm.h:368]: "An empty array" means null or a 1-element array with first item->size == 0? To be consistent with **SYS_MM_DRV_MEMORY_REGION_FOREACH**

File: tests/drivers/mm/sys_mm_drv_api/testcase.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+tests:
+  drivers.mm.sys_mm_drv_api.intel_adsp:
+    tags: drivers
+    platform_allow: intel_adsp_cavs15 intel_adsp_cavs18 intel_adsp_cavs20 intel_adsp_cavs25
```
Comment (marcinszkudlinski) [inline tests/drivers/mm/sys_mm_drv_api/testcase.yaml:4]: why those only? Looks like a generic api

Comment (edersondisouza) [reply tests/drivers/mm/sys_mm_drv_api/testcase.yaml:4]: You name it, muscle memory, copypasta, etc =P
You're right, no need to restrict testing at all.

File: include/zephyr/drivers/mm/system_mm.h (hunk 1 above)
Comment (edersondisouza) [reply include/zephyr/drivers/mm/system_mm.h:368]: Oh, empty means "containing only the sentinel at the end", will make it clearer.

**Comment by edersondisouza** [comment]
v2:
 - Don't restrict tests by platform;
 - (Try to) improve documentation about returned array of memory regions.

**Review by mwasko — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Comment by nashif** [comment]
please fix commit message

**Comment by edersondisouza** [comment]
v3: 
 - Remove WIP from commit message;
 - Rebased.

**Review by jxstelter — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by mwasko — APPROVED** [review]

**Review by marcinszkudlinski — APPROVED** [review]

**Comment by smrtos** [comment]
In future, more sophisticated tests should be added when real implementations are made.

**Pull request merged**